### ISSUE TITLE
Failure when building self-hosted NativeAOT compiler changes with 8.0 Preview 2 SDK

### ISSUE THREAD

**Issue by mthalman** [dotnet/runtime#83695]
The changes in https://github.com/dotnet/runtime/pull/81205 have caused a failure when attempting to use .NET 8 Preview 2 SDK to build the Preview 3 source. This is a required scenario for .NET's Source-Build. This was identified in the changes to update the source-build configuration so that it references the Preview 2 SDK: https://github.com/dotnet/installer/pull/15851.

The error that occurs is for the ILCompiler project:

```
EXEC : error : Failed to load assembly 'System.CommandLine' [/vmr/src/runtime/artifacts/source-build/self/src/src/coreclr/tools/aot/ILCompiler/ILCompiler.csproj]
      Internal.TypeSystem.TypeSystemException+FileNotFoundException: Failed to load assembly 'System.CommandLine'
         at Internal.TypeSystem.ThrowHelper.ThrowFileNotFoundException(ExceptionStringID, String) in /_/src/coreclr/tools/Common/TypeSystem/Common/ThrowHelper.cs:line 35
         at Internal.TypeSystem.ResolutionFailure.Throw() in /_/src/coreclr/tools/Common/TypeSystem/Common/ResolutionFailure.cs:line 105
         at Internal.TypeSystem.Ecma.EcmaModule.GetObject(EntityHandle, NotFoundBehavior ) in /_/src/coreclr/tools/Common/TypeSystem/Ecma/EcmaModule.cs:line 412
         at Internal.TypeSystem.Ecma.EcmaModule.GetType(EntityHandle) in /_/src/coreclr/tools/Common/TypeSystem/Ecma/EcmaModule.cs:line 359
         at Internal.TypeSystem.Ecma.EcmaType.InitializeBaseType() in /_/src/coreclr/tools/Common/TypeSystem/Ecma/EcmaType.cs:line 163
         at Internal.TypeSystem.Ecma.EcmaType.ComputeTypeFlags(TypeFlags) in /_/src/coreclr/tools/Common/TypeSystem/Ecma/EcmaType.cs:line 199
         at Internal.TypeSystem.TypeDesc.InitializeTypeFlags(TypeFlags) in /_/src/coreclr/tools/Common/TypeSystem/Common/TypeDesc.cs:line 135
         at Internal.TypeSystem.Ecma.EcmaType.ComputeTypeFlags(TypeFlags) in /_/src/coreclr/tools/Common/TypeSystem/Ecma/EcmaType.cs:line 247
         at Internal.TypeSystem.TypeDesc.InitializeTypeFlags(TypeFlags) in /_/src/coreclr/tools/Common/TypeSystem/Common/TypeDesc.cs:line 135
         at Internal.TypeSystem.MetadataFieldLayoutAlgorithm.ComputeInstanceLayout(DefType, InstanceLayoutKind) in /_/src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs:line 41
         at Internal.TypeSystem.DefType.ComputeInstanceLayout(InstanceLayoutKind) in /_/src/coreclr/tools/Common/TypeSystem/Common/DefType.FieldLayout.cs:line 436
         at ILCompiler.CompilerTypeSystemContext.EnsureLoadableTypeUncached(TypeDesc) in /_/src/coreclr/tools/Common/Compiler/CompilerTypeSystemContext.Validation.cs:line 137
         at Internal.TypeSystem.LockFreeReaderHashtable`2.CreateValueAndEnsureValueIsInTable(TKey) in /_/src/coreclr/tools/Common/TypeSystem/Common/Utilities/LockFreeReaderHashtable.cs:line 562
         at ILCompiler.DependencyAnalysis.EETypeNode..ctor(NodeFactory, TypeDesc) in /_/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/EETypeNode.cs:line 85
         at ILCompiler.DependencyAnalysis.NodeFactory.CreateConstructedTypeNode(TypeDesc) in /_/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/NodeFactory.cs:line 542
         at Internal.TypeSystem.LockFreeReaderHashtable`2.CreateValueAndEnsureValueIsInTable(TKey) in /_/src/coreclr/tools/Common/TypeSystem/Common/Utilities/LockFreeReaderHashtable.cs:line 562
         at ILCompiler.DependencyAnalysis.NodeFactory.ConstructedTypeSymbol(TypeDesc) in /_/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/NodeFactory.cs:line 614
         at ILCompiler.DependencyAnalysis.ReflectionInvokeMapNode.AddDependenciesDueToReflectability(DependencyList&, NodeFactory, MethodDesc) in /_/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ReflectionInvokeMapNode.cs:line 54
         at ILCompiler.MetadataManager.GetDependenciesDueToReflectability(DependencyList&, NodeFactory, MethodDesc) in /_/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/MetadataManager.cs:line 399
         at ILCompiler.DependencyAnalysis.ReflectedMethodNode.GetStaticDependencies(NodeFactory) in /_/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ReflectedMethodNode.cs:line 39
         at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.GetStaticDependenciesImpl(DependencyNodeCore`1) in /_/src/coreclr/tools/aot/ILCompiler.DependencyAnalysisFramework/DependencyAnalyzer.cs:line 182
         at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.GetStaticDependencies(DependencyNodeCore`1) in /_/src/coreclr/tools/aot/ILCompiler.DependencyAnalysisFramework/DependencyAnalyzer.cs:line 222
         at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.ProcessMarkStack() in /_/src/coreclr/tools/aot/ILCompiler.DependencyAnalysisFramework/DependencyAnalyzer.cs:line 257
         at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.ComputeMarkedNodes() in /_/src/coreclr/tools/aot/ILCompiler.DependencyAnalysisFramework/DependencyAnalyzer.cs:line 308
         at ILCompiler.ILScanner.ILCompiler.IILScanner.Scan() in /_/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ILScanner.cs:line 140
         at ILCompiler.Program.<Run>g__RunScanner|3_0(<>c__DisplayClass3_0&) in /_/src/coreclr/tools/aot/ILCompiler/Program.cs:line 438
         at ILCompiler.Program.Run() in /_/src/coreclr/tools/aot/ILCompiler/Program.cs:line 418
         at ILCompiler.ILCompilerRootCommand.<>c__DisplayClass203_0.<.ctor>b__0(InvocationContext) in /_/src/coreclr/tools/aot/ILCompiler/ILCompilerRootCommand.cs:line 272
```

An analysis of the error and its symptoms is given here: https://github.com/dotnet/installer/pull/15851#issuecomment-1474447976. As stated in the analysis, the behavior is sporadic for some reason. One notable symptom is that [`BadImageFormatException` is thrown when attempting to load certain assemblies](https://github.com/dotnet/installer/pull/15851#issuecomment-1476671469) in the `ComputeManagedAssembliesToCompileToNative` build task.

It's been confirmed that reverting the changes in https://github.com/dotnet/runtime/pull/81205 resolves the issue. We should consider reverting this change or getting a fix in to address this in the Preview 3 timeframe.

cc @MichalStrehovsky

**Comment by agocke** [comment]
> It's been confirmed that reverting the changes in https://github.com/dotnet/runtime/pull/81205 resolves the issue. We should consider reverting this change or getting a fix in to address this in the Preview 3 timeframe.

Reverting would be pretty bad -- we're planning on relying on this change to do more testing with the Pri0 coreclr tests. Otherwise the AOT compile is too slow.

How can we service the old compiler? I think we should start by getting rid of the catch and seeing the actual exception, with message.

**Comment by MichalStrehovsky** [comment]
If we need to unblock source build, adding a NativeAotSupported=false **under SourceBuild** here would unblock:

https://github.com/dotnet/runtime/blob/059898f8c0dbea2c6d0ed2ff30bc1b584af6b439/src/coreclr/tools/aot/ILCompiler/ILCompiler.csproj#L13-L14

I think this is some kind of build race caused by how source build is set up. I'm not aware of seeing this before. We do build NativeAOT compiler with NativeAOT in dotnet/runtime builds and CI and haven't seen this.

The ComputeManagedAssembliesToCompileToNative is another remnant from CoreRT/runtimelab when we could not make changes to the SDK targets to accomodate the compiler. If we address #67080 and similar by moving more of this logic into the SDK, the task can potentially be deleted. The SDK knows what assemblies we need to compile - we shouldn't have to recompute that in a task.

**Comment by mthalman** [comment]
> I think we should start by getting rid of the catch and seeing the actual exception, with message.

The exception message is `Image is too small`. Seems to come from here: https://github.com/dotnet/runtime/blob/c1acb35ce2b90f241ccdf3ead79544516c673e5f/src/libraries/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/PEBinaryReader.cs#L119

**Comment by agocke** [comment]
?? Ok this must be some sort of race condition where the binaries are still being written when we read them. Agreed with @MichalStrehovsky -- let's try to fix this authoring.

**Comment by MichaelSimons** [comment]
Gentle ping.  What is the status/eta on fixing this?  It is affecting source-build.

**Comment by agocke** [comment]
@sbomer Could you also look at this? This seems to be some incredibly strange race condition that results in a partially-written file being read by another task. From @MichalStrehovsky's comment, it sounds like we might be able to just avoid some complexity here entirely by modifying the SDK.

**Comment by MichalStrehovsky** [comment]
The ILCompiler that we're building now has a dependency on the System.CommandLine NuGet. The LKG version of ILCompiler is failing because the reference is bad.

How does source build restore the NuGets that the repo build depends on? I don't think this is fixable from the runtime repo. The problem is with the NuGet that source build is providing to the repo build system. When the task is running, the NuGets are not fully written to disk. Does source build do something with how NuGets are used? Does it run the restore as usual or there is something special going on?

**Comment by MichalStrehovsky** [comment]
Before https://github.com/dotnet/runtime/pull/81205 the partially restored nugets would still be a problem (we would likely sometimes construct a ILCompiler that doesn't work because we'd blindly copy the partial assembly) but I doubt any NativeAOT testing is happening in source build so this would never be detected.

**Comment by MichaelSimons** [comment]
> How does source build restore the NuGets that the repo build depends on?

The only intended difference between a repo build and source-build are the nuget feeds.  For source-build we have to build everything from source so any dependencies are restored from local feeds that consist of artifacts built prior in the dependency tree or from the N-1 build (necessary to break cirular dependencies)

**Comment by jkotas** [comment]
How are the contributing repo dependencies tracked by source build?

We need to guarantees that https://github.com/dotnet/dotnet/tree/main/src/command-line-api finishes building before it gets used by https://github.com/dotnet/dotnet/tree/main/src/runtime . What does guarantee that?

**Comment by MichaelSimons** [comment]
> Maybe `runtime.proj` is missing a `<RepositoryReference Include="command-line-api" />` here https://github.com/dotnet/dotnet/blob/eb24b2bf58a85e34199f319b60c30551e65984c5/repo-projects/runtime.proj#L50-L55?

I agree this dependency should be expressed but it won't actually change the build order as command-line-api is built before runtime as expressed in the [vmr's top level proj file](https://github.com/dotnet/dotnet/blob/main/repo-projects/dotnet.proj#L20).

**Comment by MichaelSimons** [comment]
Bottom line is the command-line-api is the first product repo in the build.

**Comment by jkotas** [comment]
This is regression introduced by #79783.

Before this change, the only assemblies copied next to ILCompiler.Build.Tasks.dll are System.Collections.Immutable.dll and System.Reflection.Metadata.dll.

After this change, the build is also copying over System.Buffers.dll, System.Memory.dll and System.Numerics.Vectors.dll. The problem is that these are reference assemblies that fail spectacularly when loaded for execution.

It is causing a problem in the source build environment only. It is not causing a problem in regular shipping build environment since these assemblies end up being loaded from somewhere else and so these broken reference assemblies do not come into the picture.

https://github.com/dotnet/runtime/pull/86423 is proposed fix.

**Comment by MichaelSimons** [comment]
> I have tried https://github.com/dotnet/dotnet#building a few times and it always works fine for me.

@jkotas - You must remove the workaround patch in the VMR - https://github.com/dotnet/installer/blob/main/src/SourceBuild/patches/runtime/0002-Revert-switch-to-self-hosted-NativeAOT-compiler.patch

**Comment by jkotas** [comment]
@MichaelSimons The fix is checked in. Could you please take care of deleting the workaround from source build?

**Comment by MichaelSimons** [comment]
@jkotas - I am seeing the same failure as before when trying to remove the source build workaround - https://github.com/dotnet/installer/pull/16466.  I did verify the runtime changes flowed in last week - https://github.com/dotnet/dotnet/commit/cda47539e62aa8d0cda4791d3db5a464d075b4e6#diff-531a689fcb6b7793148534b23232ba5733d6afe2ca51c1a31119defd88090bfd.

**Comment by jkotas** [comment]
@MichaelSimons Your PR was submitted against stale main that did not have the fix. I have merged current main into your PR and it is all green now.

**Comment by MichaelSimons** [comment]
> @MichaelSimons Your PR was submitted against stale main that did not have the fix. I have merged current main into your PR and it is all green now.

Thanks for updating my PR.  That confuses me though, because azdo syncs in latest when running PR validation.

**Comment by jkotas** [comment]
> azdo syncs in latest when running PR validation.

I do not think that installer repo is configured to sync when running PR validation.

**Comment by MichaelSimons** [comment]
[Source build internal CI](https://dev.azure.com/dnceng/internal/_build/results?buildId=2186521&view=results) which runs several distro variants is failing in multiple legs with the original symptoms.

**Failing Legs**
CentOSStream8_Offline_MsftSdk_x64
CentOSStream9_Offline_MsftSdk_x64
Ubuntu2004_Offline_MsftSdk_x64

Note:  There are several other legs failing but they are caused from different issues.

**Comment by jkotas** [comment]
The failure is:
```
    EXEC : error : Failed to load assembly 'System.CommandLine' [/vmr/src/runtime/artifacts/source-build/self/src/src/coreclr/tools/aot/ILCompiler/ILCompiler.csproj]
      Internal.TypeSystem.TypeSystemException+FileNotFoundException: Failed to load assembly 'System.CommandLine'
         at Internal.TypeSystem.ThrowHelper.ThrowFileNotFoundException(ExceptionStringID, String) + 0x30
         at Internal.TypeSystem.ResolutionFailure.Throw() + 0xfe
         at Internal.TypeSystem.Ecma.EcmaModule.GetObject(EntityHandle, NotFoundBehavior) + 0x9b
         at Internal.TypeSystem.Ecma.EcmaModule.GetType(EntityHandle) + 0x2c
         at Internal.TypeSystem.Ecma.EcmaType.InitializeBaseType() + 0x81
         at Internal.TypeSystem.Ecma.EcmaType.ComputeTypeFlags(TypeFlags) + 0x48
         at Internal.TypeSystem.TypeDesc.InitializeTypeFlags(TypeFlags) + 0x1e
         at Internal.TypeSystem.Ecma.EcmaType.ComputeTypeFlags(TypeFlags) + 0x1b4
         at Internal.TypeSystem.TypeDesc.InitializeTypeFlags(TypeFlags) + 0x1e
         at Internal.TypeSystem.MetadataFieldLayoutAlgorithm.ComputeInstanceLayout(DefType, InstanceLayoutKind) + 0x170
         at Internal.TypeSystem.DefType.ComputeInstanceLayout(InstanceLayoutKind) + 0x60
         at ILCompiler.CompilerTypeSystemContext.EnsureLoadableTypeUncached(TypeDesc) + 0x450
         at Internal.TypeSystem.LockFreeReaderHashtable`2.CreateValueAndEnsureValueIsInTable(TKey) + 0x14
         at ILCompiler.DependencyAnalysis.EETypeNode..ctor(NodeFactory, TypeDesc) + 0x14f
         at ILCompiler.DependencyAnalysis.NodeFactory.CreateConstructedTypeNode(TypeDesc) + 0x81
         at Internal.TypeSystem.LockFreeReaderHashtable`2.CreateValueAndEnsureValueIsInTable(TKey) + 0x14
         at ILCompiler.DependencyAnalysis.NodeFactory.ConstructedTypeSymbol(TypeDesc) + 0x77
         at ILCompiler.DependencyAnalysis.ReflectionInvokeMapNode.AddDependenciesDueToReflectability(DependencyNodeCore`1.DependencyList&, NodeFactory, MethodDesc) + 0x88
         at ILCompiler.MetadataManager.GetDependenciesDueToReflectability(DependencyNodeCore`1.DependencyList&, NodeFactory, MethodDesc) + 0x65
         at ILCompiler.DependencyAnalysis.ReflectedMethodNode.GetStaticDependencies(NodeFactory) + 0x6a
         at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.GetStaticDependenciesImpl(DependencyNodeCore`1) + 0x35
         at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.GetStaticDependencies(DependencyNodeCore`1) + 0x39
         at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.ProcessMarkStack() + 0xb1
         at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.ComputeMarkedNodes() + 0x54
         at ILCompiler.ILScanner.ILCompiler.IILScanner.Scan() + 0x19
         at ILCompiler.Program.<Run>g__RunScanner|3_0(Program.<>c__DisplayClass3_0&) + 0x117
         at ILCompiler.Program.Run() + 0x1637
         at ILCompiler.ILCompilerRootCommand.<>c__DisplayClass206_0.<.ctor>b__0(InvocationContext context) + 0x201
    /vmr/src/runtime/artifacts/source-build/self/package-cache/microsoft.dotnet.ilcompiler/8.0.0-preview.4.23259.5/build/Microsoft.NETCore.Native.targets(270,5): error MSB3073: The command ""/vmr/src/runtime/artifacts/source-build/self/package-cache/runtime.linux-x64.microsoft.dotnet.ilcompiler/8.0.0-preview.4.23259.5/tools/ilc" @"/vmr/src/runtime/artifacts/source-build/self/src/artifacts/obj/coreclr/ILCompiler/x64/Release/native/ilc.ilc.rsp"" exited with code 1. [/vmr/src/runtime/artifacts/source-build/self/src/src/coreclr/tools/aot/ILCompiler/ILCompiler.csproj]
```
Notice that this is running .NET 8.0 Preview 4 build from package-cache (package-cache/runtime.linux-x64.microsoft.dotnet.ilcompiler/8.0.0-preview.4.23259.5). This fix is present in .NET 8.0 Preview 5+ only. The failure should go away once this part of the build is upgraded to .NET 8.0 Preview 5+.

**Comment by jkotas** [comment]
The source build seems to be combining live and prebuilt binaries in a way that is not present in the regular build. Is that intentional? It is the reason why this fails in source build only while regular build is fine. Alternative way to address this issue sooner may be to make source build closer to how regular build works.

**Comment by jkotas** [comment]
> Can you point me to the indicators that source build seems to be combining live and prebuilt binaries in a way that is not present in the regular build?
> [Source build internal CI](https://dev.azure.com/dnceng/internal/_build/results?buildId=2186521&view=results) which runs several distro variants is failing in multiple legs with the original symptoms.

Notice that the `CentOSStream8_Online_MsftSdk_x64` build worked fine, but `CentOSStream8_Offline_MsftSdk_x64` build failed with the `System.CommandLine` error.

The relevant part of the [CentOSStream8_Online_MsftSdk_x64 build log](https://artprodcus3.artifacts.visualstudio.com/Ab55de4ed-4b5a-4215-a8e4-0a0a5f71e7d8/7ea9116e-9fac-403d-b258-b31fcf1bb293/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL2RuY2VuZy9wcm9qZWN0SWQvN2VhOTExNmUtOWZhYy00MDNkLWIyNTgtYjMxZmNmMWJiMjkzL2J1aWxkSWQvMjE4NjUyMS9hcnRpZmFjdE5hbWUvQ2VudE9TU3RyZWFtOF9PbmxpbmVfTXNmdFNka194NjRfQnVpbGRMb2dzX0F0dGVtcHQx0/content?format=file&subPath=%2Fsrc%2Fruntime%2Fartifacts%2FbuildLogs%2Fsource-build%2Fself%2Fsrc%2Fartifacts%2Fsourcebuild.binlog):

```
ComputeManagedAssembliesToCompileToNative
    Assembly = /vmr/src/runtime/artifacts/source-build/self/package-cache/microsoft.dotnet.ilcompiler/8.0.0-preview.4.23259.5/build/../tools/netstandard/ILCompiler.Build.Tasks.dll
    Assembly loaded during TaskRun (Build.Tasks.ComputeManagedAssembliesToCompileToNative): System.Reflection.Metadata, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a (location: /vmr/src/runtime/artifacts/source-build/self/package-cache/microsoft.dotnet.ilcompiler/8.0.0-preview.4.23259.5/tools/netstandard/System.Reflection.Metadata.dll, MVID: 245bce55-05ea-4d04-adfa-9faa17b201fd, AppDomain: [Default])
    Assembly loaded during TaskRun (Build.Tasks.ComputeManagedAssembliesToCompileToNative): System.Collections.Immutable, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a (location: /vmr/src/runtime/artifacts/source-build/self/package-cache/microsoft.dotnet.ilcompiler/8.0.0-preview.4.23259.5/tools/netstandard/System.Collections.Immutable.dll, MVID: 39e02189-e4dd-4b54-8fe2-17d6ac2c6f64, AppDomain: [Default])
```

The relevant part of the [CentOSStream8_Offline_MsftSdk_x64 build log](https://artprodcus3.artifacts.visualstudio.com/Ab55de4ed-4b5a-4215-a8e4-0a0a5f71e7d8/7ea9116e-9fac-403d-b258-b31fcf1bb293/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL2RuY2VuZy9wcm9qZWN0SWQvN2VhOTExNmUtOWZhYy00MDNkLWIyNTgtYjMxZmNmMWJiMjkzL2J1aWxkSWQvMjE4NjUyMS9hcnRpZmFjdE5hbWUvQ2VudE9TU3RyZWFtOF9PZmZsaW5lX01zZnRTZGtfeDY0X0J1aWxkTG9nc19BdHRlbXB0MQ2/content?format=file&subPath=%2Fsrc%2Fruntime%2Fartifacts%2Fsource-build%2Fself%2Fsrc%2Fartifacts%2Fsourcebuild.binlog):

```
ComputeManagedAssembliesToCompileToNative
    Assembly loaded during TaskRun (Build.Tasks.ComputeManagedAssembliesToCompileToNative): System.Reflection.Metadata, Version=8.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a (location: /vmr/src/runtime/artifacts/source-build/self/package-cache/microsoft.dotnet.ilcompiler/8.0.0-preview.4.23259.5/tools/netstandard/System.Reflection.Metadata.dll, MVID: 1c82603b-e988-482a-913f-67e66508d167, AppDomain: [Default])
    Assembly loaded during TaskRun (Build.Tasks.ComputeManagedAssembliesToCompileToNative): System.Collections.Immutable, Version=8.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a (location: /vmr/src/runtime/artifacts/source-build/self/package-cache/microsoft.dotnet.ilcompiler/8.0.0-preview.4.23259.5/tools/netstandard/System.Collections.Immutable.dll, MVID: 41f1c2d3-a8f9-4ab4-8eda-52a9000d6cf3, AppDomain: [Default])
```

The assembly version is `Version=6.0.0.0` in the successful build and `Version=8.0.0.0` in the failing build, and moreover the `Version=8.0.0.0` assembly is reference assembly instead of implementation assembly that breaks the build. Why does the offline build ends up with a completely different binary in the preview.4 bits?

**Comment by mthalman** [comment]
Ok, I've tracked down the cause of the use of [inconsistent versions](https://github.com/dotnet/runtime/issues/83695#issuecomment-1560313360). It's due to the bootstrapping logic having a casing typo for the package name. This causes two different names of the `Microsoft.DotNet.ILCompiler` package (with the same version) to exist in the local package feed. There's apparently some non-determinism in what NuGet will give you in that case which leads to the unpredictable build results. I've fixed this in my testing.

However, there is another related failure that we missed earlier through all the noise of the build errors:

```
Unhandled Exception: System.CommandLine.CommandLineException: Target OS 'centos' is not supported
  at System.CommandLine.Helpers.GetTargetOS(String) + 0x48d
  at System.CommandLine.Argument`1.<>c__DisplayClass5_0.<.ctor>b__1(ArgumentResult argumentResult, Object& value) + 0x20
  at System.CommandLine.Parsing.ArgumentResult.Convert(Argument) + 0x1c6
  at System.CommandLine.Parsing.ArgumentResult.GetArgumentConversionResult() + 0x1c
  at System.CommandLine.Parsing.ParseResultVisitor.ValidateAndConvertArgumentResult(ArgumentResult) + 0x70
  at System.CommandLine.Parsing.ParseResultVisitor.ValidateAndConvertOptionResult(OptionResult) + 0x230
  at System.CommandLine.Parsing.ParseResultVisitor.Stop() + 0x4c
  at System.CommandLine.Parsing.Parser.Parse(IReadOnlyList`1, String) + 0xdb
  at ILCompiler.Program.Main(String[] args) + 0x142
  at ilc!<BaseAddress>+0xf508d5
/tmp/MSBuildTemproot/tmpcb14709e86a14624b5eea8b5646a9ecc.exec.cmd: line 2: 27181 Aborted                 (core dumped) "/vmr/src/runtime/artifacts/source-build/self/package-cache/runtime.linux-x64.microsoft.dotnet.ilcompiler/8.0.0-preview.4.23259.5/tools/ilc" @"/vmr/src/runtime/artifacts/source-build/self/src/artifacts/obj/coreclr/ILCompiler/x64/Release/native/ilc.ilc.rsp"
```

This occurs when using a _source-built_ SDK to build the VMR. In my test scenario, the RID of the SDK is `centos.8`. This logic here gets that RID and trims it to set the value of the target OS that is passed to `ilc`:

https://github.com/dotnet/runtime/blob/10222f94e5d89b19959117dd9e5b8576ef63f878/src/coreclr/nativeaot/BuildIntegration/Microsoft.DotNet.ILCompiler.SingleEntry.targets#L7-L9

This fails in the following code because there's no corresponding case for handling a `centos` value: https://github.com/dotnet/runtime/blob/10222f94e5d89b19959117dd9e5b8576ef63f878/src/coreclr/tools/Common/CommandLineHelpers.cs#L57-L86

Here's a link to the [binlog](https://artprodcus3.artifacts.visualstudio.com/Ab55de4ed-4b5a-4215-a8e4-0a0a5f71e7d8/7ea9116e-9fac-403d-b258-b31fcf1bb293/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL2RuY2VuZy9wcm9qZWN0SWQvN2VhOTExNmUtOWZhYy00MDNkLWIyNTgtYjMxZmNmMWJiMjkzL2J1aWxkSWQvMjE5NDg4MS9hcnRpZmFjdE5hbWUvQ2VudE9TU3RyZWFtOF9PbmxpbmVfUHJldmlvdXNTb3VyY2VCdWlsdFNka194NjRfQnVpbGRMb2dzX0F0dGVtcHQx0/content?format=file&subPath=%2Fsrc%2Fruntime%2Fartifacts%2Fsource-build%2Fself%2Fsrc%2Fartifacts%2Fsourcebuild.binlog) (internal link) which comes from this [build](https://dev.azure.com/dnceng/internal/_build/results?buildId=2194881&view=logs&j=273502c4-a2ec-5aef-1ae2-d1442236a3bf&t=77157322-8c53-5d9f-77f5-d2b1faa7e281) (internal link).

@jkotas - Can you provide your input here? Clearly this code isn't comprehensive with respect to the available RIDs. But it seems broken that it should have to account for everything. Is there a better way to do this?

Because of this issue, it will delay our ability to remove the [`NativeAotSupported` patch](https://github.com/dotnet/installer/pull/15872). We need to first release an 8.0 preview version that contains a fix for this target OS issue. Then in the following preview version, we can remove the patch. They can't be done in the same release because of the need to support a workflow that allows you to build Preview N using the SDK of Preview N-1.

**Comment by agocke** [comment]
I think the RID logic in the targets is broken -- instead of using `RuntimeIdentifier` for the `_targetOS` property, that should probably be the portable RID. I forget offhand what the correct property is for that. I'll have to dig it out of the build

**Comment by jkotas** [comment]
> instead of using RuntimeIdentifier for the _targetOS property, that should probably be the portable RID. I forget offhand what the correct property is for that. I'll have to dig it out of the build

Do you mean `NETCoreSdkPortableRuntimeIdentifier` property? It is not exactly what we need here. `NETCoreSdkPortableRuntimeIdentifier` is host, but we need target. 

crossgen2 tasks have some complicate logic for this that walks RID graph here: https://github.com/dotnet/sdk/blob/dc8c847472b827dae15b3f865a5615fd7717b83e/src/Tasks/Microsoft.NET.Build.Tasks/ResolveReadyToRunCompilers.cs#L68-L73 . Is there a simpler way to do it?

> it will delay our ability to remove the https://github.com/dotnet/installer/pull/15872. We need to first release an 8.0 preview version that contains a fix for this target OS issue.

The complete removal of this patch is going to depend enabling native AOT for source build - tracked by  https://github.com/dotnet/source-build/issues/1215#issuecomment-1569474820 . If you would really like to see this patch gone, we can check in in to dotnet/runtime - we have number of similar conditions throughout the repo. It does not hurt anything.

**Comment by mthalman** [comment]
> The complete removal of this patch is going to depend enabling native AOT for source build - tracked by [dotnet/source-build#1215 (comment)](https://github.com/dotnet/source-build/issues/1215#issuecomment-1569474820) . If you would really like to see this patch gone, we can check in in to dotnet/runtime - we have number of similar conditions throughout the repo. It does not hurt anything.

I think that makes sense to do then. We're making a push to eliminate all patches for source-build.
